core/src/main/kotlin/com/strumenta/kolasu/traversing/Structurally.kt

```diff
@@ -153,3 +165,33 @@ fun <T> Node.searchByType(
 fun <T> Node.collectByType(klass: Class<T>, walker: KFunction1<Node, Sequence<Node>> = Node::walk): List<T> {
     return walker.invoke(this).filterIsInstance(klass).toList()
 }
+
+/**
+ * The FastWalker is a walker that implements a cache to speed up subsequent walks.
+ * The first walk will take the same time of a normal walk.
+ * This walker will ignore any change to the nodes.
+ */
+class FastWalker(val node: Node) {
+    private val childrenMap: WeakHashMap<Node, List<Node>> = WeakHashMap<Node, List<Node>>()
+
+    private fun getChildren(child: Node): List<Node> {
+        return if (childrenMap.containsKey(child)) {
+            childrenMap[child]!!
+        } else {
+            childrenMap.put(child, child.walkChildren().toList())
+            childrenMap[child]!!
+        }
+    }
+    fun walk(root: Node = node): Sequence<Node> {
```

Comment:
This works, but it duplicates the walking algorithm, and constrains the user to this single walk algorithm (so e.g. no `walkLeavesFirst`, `walkAncestors` and so on).

I was thinking about adding an extension method to `Sequence<Node>` like:

```
fun <T: Node> Sequence<T>.cached(): Sequence<T> {
    return this.toCollection(mutableListOf()).asSequence()
}
```

Then, one could do e.g. 

```
val nodes = root.walkLeavesFirst().cached()
nodes.forEach { ... }
// ...
nodes.map { ... }
```

so you pay the cost of the traversal only once. With my implementation, you pay it eagerly when calling `cached()`, probably there's a nice way to make it lazy so you only pay it the first time you actually traverse the nodes.

Reply:
If I understand it correctly, you are saying that the FastWalker should look more like this:

```
class FastWalker(val node: Node) {
    private val walkMap: WeakHashMap<Node, Sequence<Node>> = WeakHashMap<Node, Sequence<Node>>()


    fun walk(root: Node = node): Sequence<Node> {      
        return if (walkMap.containsKey(node)) {
            walkMap[node]!!
        } else {
            walkMap.put(node, walk(node).cached())
            walkMap[node]!!
        }
    }
}
```

Is that correct?

Reply:
If you are saying that FastWalker itself is a bit superflous, I agree that it is kinda true. Its purpose is essentially to cover a hidden drawback of kolasu. The FastWalker is simply a convenience for people that are not aware that the walk takes more time than expected because it uses reflection. So it essentially automates saving the result of the walk for the user. If you were aware of the issue, you could already achieve the same results manually, right now, simply by saving the result of the walk.

So, what I mean is that I do not think that this:

`val nodes = root.walk().cached()`

would do something different than this, right now:

`val nodes = root.walk()`

At least `FastWalker` does also speed up subsequent walks starting at other points of the tree.

Reply:
Well, using the FastWalker is still an explicit choice, so you have to be aware of the problem anyway.

Your other point makes a real difference. If the user's pattern is to walk the same tree repeatedly starting from different nodes, then something like FastWalker is necessary to speed up the traversal. To avoid having to code a different FastWalker for each walking strategy, we could re-engineer FastWalker to work on sequences rather than nodes, something like the following:

```
class FastWalker(val sequence: Sequence<Node>) {
    private val walks = mutableMapOf<Node, Sequence<Node>>()

    fun walk(root: Node): Sequence<Node> {
        return walks.computeIfAbsent(root) {
            val iterator = sequence.iterator()
            for (node in iterator) {
                if (node == root) {
                    val cached = iterator.asSequence().toList()
                    return@computeIfAbsent sequence {
                        yield(node)
                        yieldAll(cached)
                    }
                }
            }
            throw Error("Node not in sequence")
        }
    }
}
```

I've checked that the above passes your tests. You use it like this:

```
val fw = FastWalker(root.walk())
fw.walk(root).forEach { countedNodes++ } // Slow
fw.walk(root).forEach { countedNodes++ } // Fast
```

Reply:
Ah, but this would still be slow the first time your walk starting from another root, it only improves if you walk twice from the same root.

Reply:
So, I think we should keep FastWalker as it is, and document that if you want a different traversal algorithm than `walk()`, you'll have to code your own version of FastWalker.